Source/DafnyCore/AST/TopLevelDeclarations.cs

```diff
@@ -845,6 +845,21 @@ public class ModuleDefinition : IDeclarationOrUsage, INamedRegion, IAttributeBea
     }
   }
 
+  /// <summary>
+  /// Yields all functions and methods that are members of some type in the given list of
+  /// declarations, including prefix lemmas and prefix predicates.
+  /// </summary>
+  public static IEnumerable<ICallable> AllCallablesIncludingPrefixDeclarations(List<TopLevelDecl> declarations) {
```

Comment:
I'd like to have each language construct be as independent as possible. For extreme lemmas and predicates this would also mean they have mostly outwards references. Would it be possible to achieve this?

Would it be possible to translate extremes to their prefix versions and then forget about the originals? Or maybe create the prefix versions next to the originals?

Reply:
We need both. A program needs the extremes, since that's what you'll ultimately will use for the property you want. The prefix predicates/lemmas give a way to prove properties about the extremes, so they are also needed.

Actually, the more surprising thing (to me) is that `AllCallables` (which excludes the prefix versions) is used so many times.

Reply:
> We need both. A program needs the extremes, since that's what you'll ultimately will use for the property you want. The prefix predicates/lemmas give a way to prove properties about the extremes, so they are also needed.

Sure, so can't we place both side by side?

Reply:
I don't understand. The `AllCallablesIncludingPrefixDeclarations` and `AllCallables` properties are declared side by side. Are you perhaps suggesting that the prefix predicates/lemmas be placed in the list of the enclosing module's top-level-decls (rather than being linked off the corresponding extreme predicate/lemma)? (And if so, do you think that will improve the organization for other parts of the implementation?)

Reply:
I looked into this in detail for `.PrefixLemma` and `.PrefixPredicate`. In places where the code looked like your first example, I change `AllCallables` to `AllCallablesIncludingPrefixDeclarations`. I also improved code in the various other visitors, making use of newer C# constructs. And I improved the iteration over declarations in `IsEssentiallyEmptyModuleBody()`, to be more exact than before.

In the code base, what happens with extreme lemmas/predicates and prefix lemmas/predicates is quite varied. In some places, the extreme declaration is used and the prefix version ignored. In other places, it's the other way around. And in some places, the two are handled in different ways. Therefore, I suggest that any further improvements be done separately from this PR.